flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/TimeAttributeITCase.scala

```diff
@@ -0,0 +1,162 @@
+  TestCollectionTableFactory.isStreaming = true
+
+  @Test
+  def testWindowAggregateOnWatermark(): Unit = {
+    val ddl =
+      """
+        |CREATE TABLE src (
+        |  ts TIMESTAMP(3),
+        |  a INT,
+        |  b DOUBLE,
+        |  WATERMARK FOR ts AS ts - INTERVAL '0.001' SECOND
+        |) WITH (
+        |  'connector' = 'COLLECTION'
+        |)
+      """.stripMargin
+    val query =
+      """
+        |SELECT TUMBLE_END(ts, INTERVAL '0.003' SECOND), COUNT(ts), SUM(b)
+        |FROM src
+        |GROUP BY TUMBLE(ts, INTERVAL '0.003' SECOND)
+      """.stripMargin
+    tEnv.sqlUpdate(ddl)
+    val sink = new TestingAppendSink()
+    tEnv.sqlQuery(query).toAppendStream[Row].addSink(sink)
+    tEnv.execute("SQL JOB")
+
+    val expected = Seq(
+      "1970-01-01T00:00:00.003,2,3.0",
+      "1970-01-01T00:00:00.006,2,7.0",
+      "1970-01-01T00:00:00.009,2,6.0",
+      "1970-01-01T00:00:00.018,1,4.0")
+    assertEquals(expected.sorted, sink.getAppendResults.sorted)
+  }
+
+  @Test
+  def testWindowAggregateOnCustomizedWatermark(): Unit = {
+    JavaFunc5.openCalled = false
+    JavaFunc5.closeCalled = false
+    tEnv.registerFunction("myFunc", new JavaFunc5)
+    val ddl =
+      """
+        |CREATE TABLE src (
+        |  ts TIMESTAMP(3),
+        |  a INT,
+        |  b DOUBLE,
+        |  WATERMARK FOR ts AS myFunc(ts, a)
+        |) WITH (
+        |  'connector' = 'COLLECTION'
+        |)
+      """.stripMargin
+    val query =
+      """
+        |SELECT TUMBLE_END(ts, INTERVAL '0.003' SECOND), COUNT(ts), SUM(b)
+        |FROM src
+        |GROUP BY TUMBLE(ts, INTERVAL '0.003' SECOND)
+      """.stripMargin
+    tEnv.sqlUpdate(ddl)
+    val sink = new TestingAppendSink()
+    tEnv.sqlQuery(query).toAppendStream[Row].addSink(sink)
+    tEnv.execute("SQL JOB")
+
+    val expected = Seq(
+      "1970-01-01T00:00:00.003,2,3.0",
+      "1970-01-01T00:00:00.006,2,7.0",
+      "1970-01-01T00:00:00.009,2,6.0",
+      "1970-01-01T00:00:00.018,1,4.0")
+    assertEquals(expected.sorted, sink.getAppendResults.sorted)
+    assertTrue(JavaFunc5.openCalled)
+    assertTrue(JavaFunc5.closeCalled)
+  }
+
+  @Test
+  def testWindowAggregateOnNestedRowtime(): Unit = {
+    val ddl =
+      """
+        |CREATE TABLE src (
+        |  col ROW<
+        |    ts TIMESTAMP(3),
+        |    a INT,
+        |    b DOUBLE>,
+        |  WATERMARK FOR col.ts AS col.ts - INTERVAL '0.001' SECOND
+        |) WITH (
+        |  'connector' = 'COLLECTION'
+        |)
+      """.stripMargin
+    val query =
+      """
+        |SELECT TUMBLE_END(col.ts, INTERVAL '0.003' SECOND), COUNT(*)
+        |FROM src
+        |GROUP BY TUMBLE(col.ts, INTERVAL '0.003' SECOND)
+      """.stripMargin
+    tEnv.sqlUpdate(ddl)
+    expectedException.expect(classOf[TableException])
+    expectedException.expectMessage(
+      "Nested field 'col.ts' as rowtime attribute is not supported right now.")
+    tEnv.sqlQuery(query)
+  }
+
```

Comment:
Also add a test for multiple watermark definitions.

Reply:
We already have a test in sql parser `FlinkSqlParserImplTest#testCreateTableWithMultipleWatermark`.